source/common/conn_pool/conn_pool_base.cc

```diff
@@ -600,6 +609,28 @@ void ActiveClient::onConnectTimeout() {
   close();
 }
 
+void ActiveClient::onLifetimeTimeout() {
+  // The lifetime timer should only have started after we left the CONNECTING state.
+  ENVOY_BUG(state_ != ActiveClient::State::CONNECTING, "lifetime timeout while connecting");
+
+  // There's nothing to do if the client is connecting, closed or draining.
+  if (state_ == ActiveClient::State::CONNECTING || state_ == ActiveClient::State::CLOSED ||
```

Comment:
Very nice.  I had a long comment about how there was a lifetime bug because I missed this, until I saw that you tested it already and went back to figure out why it wasn't a problem :-)

optionally, I would suggest disarming the timer in onConnectionEvent where we disable the connection timer.  It feels safer to not have an alarm fire once the session is on the deferred delete list, WDYT?

Reply:
Yes it would definitely be safest to explicitly disable the timer in that case. I'll take a look at that.